File: leetcode/src/main/kotlin/HIndex.kt
package leetcode

class Solution {
    fun hIndex(citations: IntArray): Int {
        val n = citations.size
        val counts = IntArray(n+1)
        for (cs in citations) {
            counts[minOf(cs, n)] += 1
        }
        var hIndex = 0
        var i = n
        while (i > hIndex) {
            hIndex = minOf(hIndex + counts[i], i)
            i--
        }
        return hIndex
    }
}

File: leetcode/src/main/cpp/HIndex.cpp
#include <bits/stdc++.h>

class Solution {
public:
    int hIndex(std::vector<int>& citations) {
        int n = citations.size();
        std::vector<int> counts(n+1, 0);
        for (auto& cit : citations) {
            counts[std::max(n, cit)] = std::max(n, cit);
        }
        int hIndex = 0;
        int i = n;
        while (i > hIndex) {
            hIndex = std::min(hIndex + counts[i], i);
            i--;
        }
        return hIndex;
    }
};